docs/RationaleHWArith.md

```diff
@@ -0,0 +1,73 @@
+
+In general, the bit width inference rules are designed to model SystemVerilog semantics. However, unlike SystemVerilog, the aim of this dialect is to provide a strongly-typed hardware arithmetic library with the width inference rules for every operator explained in a clear and understandable way to avoid ambiguity.
+
+To serve this goal of being strongly typed, the operations of the dialect explicitly do not accept signless types. As such, only `ui#`, `si#` types are allowed. A casting operation is provided to serve signedness casting as well as truncation and padding.
+
+Below we try to capture some common questions on the capabilities of such a dialect, and how we've addressed/think about the issue.
+
+### Q&A
+* **Q:** Does this support n-ary operations?
+    * n-ary operations in this dialect would have to solve the issue of how width rules in expressions such as the following are handled:
+    `%0 = hwarith.add %0, %1, %2 : si6, si5, si4`  
+    In the short term, we envision this dialect to be considering strictly binary operations. Defining width inference rules for these is simpler and fits the immediate usecase. If, in the future, a need for n-ary operations comes up and is motivated clearly (e.g. for optimization purposes), the dialect should be adapted to support it. In this case, we expect n-ary operation rules to be a superset of those defined for binary operations.
+* **Q:** Does this support 0-width operands?
+    * 0-width values might arise from arithmetic rules which reduce the bit width of an expression wrt. the operands to that expression. One case where such rules _may_ apply is in the implementation of modulo operations.  
+    We refrain from adding support at this point in time, since support for 0-width values in the remainder of the RTL dialects is, at time of writing, lacking/undefined. Once support is added in these downstream dialects, 0-width support in `hwarith` should be reconsidered.
+* **Q:** Does this support width inferred types?
+    * Relying on width inference is relevant when referencing results of other width-inferred value. Without this, a user/frontend must itself know and apply width inference rules before generating the IR. Having width-inferred types will be convenient not only for generating IR, but also to leave room for width inference rules and optimizations to apply recursively. As an example:
+    ```mlir
+    %1 = hwarith.add %a, %b : ui4, ui4
+    %2 = hwarith.add %c, %1 : ui3, ui // %1 is an unsigned integer of inferred width
+    ```
+    We see merit in having inferred width types, but recognize that the work required to get this right is non-trivial and will require significant effort. Once need arises, and resources are available, we find width inferred types to be a valuable contribution to the dialect, but it will not be part of the initial implementation.
```

Comment:
What then is being added?  essentially operations with implicit extensions?

Reply:
The term "inferred width types" in this answer is referring to not having to specify the width(s) of the operands, not the result type. This dialect overall is computing extended width results a la SV/VHDL.

@mortbopet maybe it's best to not elide the result types in your examples throughout this document so as to be clear as to what is going on.

Reply:
Result types are now intended to be explicit (through function-like typing).